internal/bootloader/grubenv/grubenv.go

```diff
@@ -0,0 +1,117 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package grubenv
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/canonical/x-go/strutil"
+)
+
+// FIXME: support for escaping (embedded \n in grubenv) missing
+type Env struct {
+	env      map[string]string
+	ordering []string
+
+	path string
+}
+
+func NewEnv(path string) *Env {
+	return &Env{
+		env:  make(map[string]string),
+		path: path,
+	}
+}
+
+func (g *Env) Get(name string) string {
+	return g.env[name]
+}
+
+func (g *Env) Set(key, value string) {
+	if !strutil.ListContains(g.ordering, key) {
+		g.ordering = append(g.ordering, key)
+	}
+
+	g.env[key] = value
+}
+
+func (g *Env) Load() error {
+	buf, err := ioutil.ReadFile(g.path)
+	if err != nil {
+		return err
+	}
+	if len(buf) != 1024 {
+		return fmt.Errorf("grubenv %q must be exactly 1024 byte, got %d", g.path, len(buf))
+	}
+	if !bytes.HasPrefix(buf, []byte("# GRUB Environment Block\n")) {
+		return fmt.Errorf("cannot find grubenv header in %q", g.path)
+	}
+	rawEnv := bytes.Split(buf, []byte("\n"))
+	for _, env := range rawEnv[1:] {
+		l := bytes.SplitN(env, []byte("="), 2)
+		// be liberal in what you accept
```

Comment:
I think the GRUB file can have other `# comment` lines in it (mine does), and there could be an `=` sign in a comment. So instead of looping over `rawEnv` (which should probably be called `lines`), how about we loop over all the lines, and `if bytes.HasPrefix(line, []byte("#")) { continue }`?